TypeError: Cannot read property 'fileName' of undefined, in v3.0.13

Hi..
I'm trying to code coverage my angular (typescript) application with karma, karma-typescript, etc... but I got an error message. Could you help me please ?

Error message:
```
C:\Personal_Data\Web-Client>yarn test
yarn run v1.7.0
$ ng test
 10% building modules 1/1 modules 0 active(node:2280) DeprecationWarning: Tapable.plugin is deprecated. Use new API on `.hooks` instead
 13% building modules 27/28 modules 1 active …Client\src\app\login\auth.guard.spec.ts27 07 2018 14:22:44.314:INFO [compiler.karma-typescript]: Compiling project using Typescript 2.7.2
27 07 2018 14:22:49.546:INFO [compiler.karma-typescript]: Compiled 73 files in 5187 ms.
27 07 2018 14:22:59.645:INFO [compiler.karma-typescript]: Compiling project using Typescript 2.7.2
27 07 2018 14:23:02.961:INFO [compiler.karma-typescript]: Compiled 73 files in 3301 ms.
27 07 2018 14:23:03.206:INFO [bundler.karma-typescript]: Bundled imports for 130 file(s) in 13150 ms.
27 07 2018 14:23:04.041:ERROR [karma]: TypeError: Cannot read property 'filename' of undefined
    at Bundler.addLoaderFunction (C:\Personal_Data\Web-Client\node_modules\karma-typescript\src\bundler\bundler.ts:172:77)
    at C:\Personal_Data\Web-Client\node_modules\karma-typescript\src\bundler\bundler.ts:158:42
    at Array.forEach (<anonymous>)
    at C:\Personal_Data\Web-Client\node_modules\karma-typescript\src\bundler\bundler.ts:157:34
    at C:\Personal_Data\Web-Client\node_modules\karma-typescript\src\bundler\bundler.ts:242:13
    at C:\Personal_Data\Web-Client\node_modules\graceful-fs\graceful-fs.js:43:10
    at FSReqWrap.oncomplete (fs.js:152:20)
(node:2280) [DEP0079] DeprecationWarning: Custom inspection function on Objects via .inspect() is deprecated
Done in 31.28s.
```

Here are my configs:
**package.json**
```
{
  "name": "mywebclient",
  "version": "0.0.0",
  "license": "MIT",
  "scripts": {
    "ng": "ng",
    "start": "ng serve",
    "build": "ng build",
    "test": "ng test",
    "lint": "ng lint --type-check",
    "e2e": "ng e2e",
    "postinstall": "yarn run update_markdown %NODE_ENV%",
    "update_markdown": "node update.markdown.js",
    "update": "node update.packages.js",
    "updateAll": "node update.packages.js --all"
  },
  "private": true,
  "dependencies": {
    "@angular/animations": "6.0.9",
    "@angular/cdk": "6.4.1",
    "@angular/common": "6.0.9",
    "@angular/compiler": "6.0.9",
    "@angular/core": "6.0.9",
    "@angular/flex-layout": "6.0.0-beta.16",
    "@angular/forms": "6.0.9",
    "@angular/http": "6.0.9",
    "@angular/material": "6.4.1",
    "@angular/platform-browser": "6.0.9",
    "@angular/platform-browser-dynamic": "6.0.9",
    "@angular/router": "6.0.9",
    "bootstrap": "4.1.1",
    "core-js": "2.5.7",
    "font-awesome": "4.7.0",
    "hammerjs": "2.0.8",
    "material-design-icons": "3.0.1",
    "ng-event-source": "1.0.11",
    "rxjs": "6.2.2",
    "typedoc": "0.11.1",
    "typescript": "2.7.2",
    "zone.js": "0.8.26"
  },
  "devDependencies": {
    "@angular-devkit/build-angular": "0.6.8",
    "@angular/cli": "6.0.8",
    "@angular/compiler-cli": "6.0.9",
    "@angular/language-service": "6.0.9",
    "@types/jasmine": "2.8.8",
    "@types/jasminewd2": "2.0.3",
    "@types/node": "10.5.2",
    "codelyzer": "4.4.2",
    "jasmine-core": "3.1.0",
    "jasmine-spec-reporter": "4.2.1",
    "karma": "2.0.4",
    "karma-chrome-launcher": "2.2.0",
    "karma-cli": "1.0.1",
    "karma-coverage": "1.1.2",
    "karma-coverage-istanbul-reporter": "2.0.1",
    "karma-firefox-launcher": "1.1.0",
    "karma-jasmine": "1.1.2",
    "karma-jasmine-html-reporter": "1.2.0",
    "karma-json-reporter": "1.2.1",
    "karma-junit-reporter": "1.2.0",
    "karma-typescript": "3.0.13",
    "karma-typescript-es6-transform": "1.0.4",
    "protractor": "5.4.0",
    "ts-node": "7.0.0",
    "tslint": "5.11.0"
  }
}
```
**karma.conf.js**
```
// Karma configuration file, see link for more information
// https://karma-runner.github.io/1.0/config/configuration-file.html

module.exports = function (config) {
  config.set({
    basePath: '',
    frameworks: ['jasmine', '@angular-devkit/build-angular', 'karma-typescript'],
    plugins: [
      require('karma-jasmine'),
      require('karma-chrome-launcher'),
      require('karma-jasmine-html-reporter'),
      require('karma-coverage-istanbul-reporter'),
      require('@angular-devkit/build-angular/plugins/karma'),
      require('karma-junit-reporter'),
      require('karma-typescript'),
      require('karma-typescript-es6-transform')
    ],
    client:{
      clearContext: false // leave Jasmine Spec Runner output visible in browser
    },
    karmaTypescriptConfig: {
      bundlerOptions: {
        transforms: [
          require('karma-typescript-es6-transform')()
        ]
      },
      reports: {
        'cobertura': {
          'directory': 'tests_results/coverage',
          'subdirectory': 'cobertura',
          'filename': 'coverage.xml'
        }
      },
      tsconfig: './../tsconfig.json'
    },
    files: ['app/**/*.ts'],
    preprocessors: {
      'app/**/*.ts': 'karma-typescript'
    },
    reporters: ['dots', 'junit', 'karma-typescript'],
    port: 9876,
    colors: true,
    logLevel: config.LOG_INFO,
    autoWatch: true,
    browsers: ['Chrome'],
    singleRun: true,
    junitReporter: {
      outputFile: 'test-results.xml',
      outputDir: 'tests_results'
    },
  });
};
```
**tsconfig.json**
```
{
  "compileOnSave": false,
  "compilerOptions": {
    "baseUrl": "./",
    "outDir": "./dist/out-tsc",
    "sourceMap": true,
    "declaration": false,
    "moduleResolution": "node",
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "target": "es5",
    "typeRoots": [
      "node_modules/@types"
    ],
    "lib": [
      "es2017",
      "dom"
    ]
  }
}
```

I think it happens when bundling takes >500 ms (BUNDLE_DELAY in bundler.ts):

```
[2018-09-19 00:06:13.621] [INFO] bundler.karma-typescript - Bundled imports for 8 file(s) in 616 ms.
[2018-09-19T00:06:13.623] [ERROR] karma - TypeError: Cannot read property 'filename' of undefined
```

Nope, getting it for <500 ms also:

```
[2018-09-19 10:46:38.559] [INFO] bundler.karma-typescript - Bundled imports for 8 file(s) in 396 ms.
[2018-09-19T10:46:38.560] [ERROR] karma - TypeError: Cannot read property 'filename' of undefined
```

--- a/packages/karma-typescript/src/bundler/bundler.ts
+++ b/packages/karma-typescript/src/bundler/bundler.ts
@@ -73,11 +73,15 @@ export class Bundler {
     }
 
     private bundleQueuedModules() {
+        // Drain the bundle queue and keep a copy of the current items. This prevents problems
+        // when a new bundle is queued while transformations are running.
+        const queuedItems = this.bundleQueue.slice(0);
+        this.bundleQueue.length = 0;
 
         const benchmark = new Benchmark();
 
-        this.transformer.applyTsTransforms(this.bundleQueue, () => {
-            this.bundleQueue.forEach((queued) => {
+        this.transformer.applyTsTransforms(queuedItems, () => {
+            queuedItems.forEach((queued) => {
 
                 const source = this.sourceMap.removeSourceMapComment(queued);
                 const map = this.sourceMap.getSourceMap(queued);
@@ -91,13 +95,13 @@ export class Bundler {
                     queued.file.path, queued.file.originalPath, source, map);
             });
 
-            const dependencyCount = this.dependencyWalker.collectTypescriptDependencies(this.bundleQueue);
+            const dependencyCount = this.dependencyWalker.collectTypescriptDependencies(queuedItems);
 
             if (this.shouldBundle(dependencyCount)) {
-                this.bundleWithLoader(benchmark);
+                this.bundleWithLoader(benchmark, queuedItems);
             }
             else {
-                this.bundleWithoutLoader();
+                this.bundleWithoutLoader(queuedItems);
             }
         });
     }
@@ -119,9 +123,9 @@ export class Bundler {
         return this.projectImportCountOnFirstRun > 0;
     }
 
-    private bundleWithLoader(benchmark: Benchmark) {
+    private bundleWithLoader(benchmark: Benchmark, queuedItems: Queued[]) {
 
-        async.each(this.bundleQueue, (queued, onQueuedResolved) => {
+        async.each(queuedItems, (queued, onQueuedResolved) => {
 
             this.addEntrypointFilename(queued.item.filename);
 
@@ -131,14 +135,14 @@ export class Bundler {
                 });
             }, onQueuedResolved);
         }, () => {
-            this.onAllResolved(benchmark);
+            this.onAllResolved(benchmark, queuedItems);
         });
     }
 
-    private bundleWithoutLoader() {
+    private bundleWithoutLoader(queuedItems: Queued[]) {
         this.globals.add(this.bundleBuffer, this.entrypoints, () => {
             this.writeMainBundleFile(() => {
-                this.bundleQueue.forEach((queued) => {
+                queuedItems.forEach((queued) => {
                     const source = queued.item.source + "\n" +
                         (queued.item.sourceMap ? queued.item.sourceMap.toComment() + "\n" : "");
                     queued.callback(source);
@@ -147,23 +151,21 @@ export class Bundler {
         });
     }
 
-    private onAllResolved(benchmark: Benchmark) {
+    private onAllResolved(benchmark: Benchmark, queuedItems: Queued[]) {
 
         this.orderEntrypoints();
 
         this.globals.add(this.bundleBuffer, this.entrypoints, () => {
             this.writeMainBundleFile(() => {
                 this.log.info("Bundled imports for %s file(s) in %s ms.",
-                    this.bundleQueue.length, benchmark.elapsed());
+                    queuedItems.length, benchmark.elapsed());
 
-                this.bundleQueue.forEach((queued) => {
+                queuedItems.forEach((queued) => {
                     queued.callback(this.addLoaderFunction(queued.item, true));
                 });
 
                 this.log.debug("Karma callbacks for %s file(s) in %s ms.",
-                    this.bundleQueue.length, benchmark.elapsed());
-
-                this.bundleQueue.length = 0;
+                    queuedItems.length, benchmark.elapsed());
             });
         });
     }

